evennia/web/webclient/static/webclient/js/webclient_gui.js

```diff
@@ -76,36 +96,91 @@ function doSendText() {
 // catch all keyboard input, handle special chars
 function onKeydown (event) {
     var code = event.which;
-    inputfield = $("#inputfield");
+    var history_entry = null;
+    var inputfield = $("#inputfield");
     inputfield.focus();
 
     if (code === 13) { // Enter key sends text
         doSendText();
         event.preventDefault();
-    }
-    else if (code === 38) { // Arrow up
-        inputfield.val(input_history.back());
+    } 
+    else if (inputfield[0].selectionStart == inputfield.val().length) { 
+        // Only process up/down arrow if cursor is at the end of the line.
+        if (code === 38) { // Arrow up
+            history_entry = input_history.back();
+        }
+        else if (code === 40) { // Arrow down
+            history_entry = input_history.fwd();
+        }
+    } 
+    
+    if (history_entry !== null) {
+        // Doing a history navigation; replace the text in the input.
+        inputfield.val(history_entry);
         event.preventDefault();
     }
-    else if (code === 40) { // Arrow down
-        inputfield.val(input_history.fwd());
-        event.preventDefault();
+    else {
+        // Save the current contents of the input to the history scratch area.
+        setTimeout(function () {
+            // Need to wait until after the key-up to capture the value.
+            input_history.scratch(inputfield.val());
+            input_history.end();
+        }, 0);
     }
 };
 
+var resizeInputField = function () {
+    var min_height = 50;
+    var max_height = 300;
+    var prev_text_len = 0;
+    
+    // Check to see if we should change the height of the input area
+    return function () {
+        var inputfield = $("#inputfield");
+        var scrollh = inputfield.prop("scrollHeight");
+        var clienth = inputfield.prop("clientHeight");
+        var newh = 0;
+        var curr_text_len = inputfield.val().length;
+        
+        if (scrollh > clienth && scrollh <= max_height) {
+            // Need to make it bigger
+            newh = scrollh;
+        } 
+        else if (curr_text_len < prev_text_len) {
```

Comment:
That I'm aware of, yeah. The only alternate solution I can think of would be to figure out the metrics for whatever font is being used and then use the character count, but this would get hairy and wouldn't account for line-height. I'll make the CSS a bit more consistent; at the moment with the current set of CSS rules, it works fine as written (the extra rules don't affect the dimensions of the sizer or textarea differently), but it'd be too easy to slip a rule in later to throw things off as it stands, so I'll fix that.